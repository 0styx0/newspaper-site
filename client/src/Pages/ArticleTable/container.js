import React from 'react';
import fetchFromApi from '../../helpers/fetchFromApi';

import ArticleTable from './';

export default class ArticleTableContainer extends React.Component {

    constructor() {
        super();

        this.putData = this.putData.bind(this);

        this.state = {
            history: []
        };
    }

    async putData(num = null) {

        const lastPath = +window.location.pathname.split("/").slice(-1);

        num = (isNaN(lastPath) || ((!isNaN(num)) && num)) ? num : lastPath;


        // use cached results (don't load new info)
        if (this.state.history[num]) {
            this.setState({
                articles: this.state.history[num].articles,
                issueInfo: this.state.history[num].issueInfo
            });

            window.history.pushState({}, `Issue ${num}`, num);
        }
        else {

            const data = await this.getData(num);
            const json = await data.json();

            window.history.pushState({}, `Issue ${num}`, isNaN(lastPath) ? "modifyArticles/"+json[2].num : json[2].num);

            this.setArticleInfo(json);
        }
    }

    componentWillMount() {

        this.putData(null);
    }

    async getData(num) {

         return await fetchFromApi(`articleGroup?articlesFor=${num}`);

    }


    setArticleInfo(data) {

        const history = [...this.state.history];
        history[data[2].num] = {articles: data[0], issueInfo: data[2]};

        this.setState({
            articles: data[0].length ? data[0] :  null,
            issueInfo: data[2],
            history
        });
    }

    render() {

            return <ArticleTable
                     articles={this.state.articles}
                     issue={this.state.issueInfo}
                     update={(e) => this.putData(e.target.value)}
                   />
    }
}
