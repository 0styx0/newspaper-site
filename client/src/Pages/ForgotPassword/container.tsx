import * as React from 'react';
import { graphql, withApollo } from 'react-apollo';
import { RecoverPassword } from '../../graphql/user';
import ForgotPassword from './';

interface Props {
    recoverPassword:
      (params: {
          query: typeof RecoverPassword,
          variables: { username: string, authCode: string, email: string }
        }
      ) => Promise<{message: string}>;
}

export class ForgotPasswordContainer extends React.Component<Props, {}> {

    constructor(props: Props) {
        super(props);

        this.onSubmit = this.onSubmit.bind(this);

    }

    /**
     * Sends data to server
     */
    onSubmit(target: HTMLFormElement) {

        this.props.recoverPassword({
            query: RecoverPassword,
            variables: {
                authCode: (target.querySelector('input[name=lastAuth]') as HTMLInputElement).value,
                username: (target.querySelector('input[name=username]') as HTMLInputElement).value,
                email: (target.querySelector('input[name=email]') as HTMLInputElement).value,
            }
        });
    }

    render() {
        return <ForgotPassword onSubmit={this.onSubmit} />;
    }
}

const ForgotPasswordContainerWithData =
    graphql(RecoverPassword, {name: 'recoverPassword'})(ForgotPasswordContainer as any);

export default withApollo(ForgotPasswordContainerWithData);
