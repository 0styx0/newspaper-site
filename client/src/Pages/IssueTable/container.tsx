import * as React from 'react';
import { compose, graphql } from 'react-apollo';
import { IssueQuery, IssueUpdate } from '../../graphql/issues';
import { Issue } from './interface.shared';
import IssueTable from './';

interface State {
    privateIssue: { // admins can change these (until public is true)
        public?: boolean;
        name?: string;
    };
    loaded: boolean;
}

interface Props {
    data: {
        loading: boolean;
        issues: Issue[]
    };
    mutate: Function;
}

/**
 * Creates a table of issues, in descending order with info of
 * number, name, datePublished, and views (sum of views of all articles in the issue)
 *
 * If user is admin (lvl 3) and issue is not public, user can change issue name and make it public.
 * In that case, @see IssueTable.allowEditsOfLastIssue for difference in UI
 */
export class IssueTableContainer extends React.Component<Props, State> {

    constructor() {
        super();

        this.onChangeIssueInfo = this.onChangeIssueInfo.bind(this);
        this.onSubmit = this.onSubmit.bind(this);

        this.state = {
            privateIssue: {},
            loaded: false
        };
    }

    /**
     * When graphql gets data from @see graphql/issues,
     * this turns it into 2d array of what will be table rows and saves it to state.issueInfo
     *
     * If current user is an admin, enables editing of most recent, unpublished issue @see this.allowEditsOfLastIssue
     */
    componentWillReceiveProps(props: Props) {

        if (!props.data.issues || this.props.data.issues.length > 0) {
            return;
        }

        this.setState({loaded: true});
    }

    /**
     * @param e {HTMLInputElement event}
     *
     * Adds input's value to this.state.privateIssue[input.name]
     *
     * @example if event.target = <input name="public" value="1" />, then after this, this.state.privateIssue.public = 1
     */
    onChangeIssueInfo(e: Event) {

        const target = e.target as HTMLInputElement;

        const privateIssue = Object.assign({}, this.state.privateIssue);
        privateIssue[target.name] = target.value;

        this.setState({
            privateIssue
        });
    }

    /**
     * @param e {HTMLFormElement event}
     *
     * Sends this.state.privateIssue data to server (so name and/or public status can be saved to db)
     */
    onSubmit(e: Event) {

        e.stopPropagation();
        e.preventDefault();

        this.props.mutate({
            variables: {
                name: this.state.privateIssue.name,
                public: !!this.state.privateIssue.public
            }
        });
    }

    render() {

        if (this.props.data.issues.length < 1) {
            return null;
        }

        return (
            <IssueTable
              key={+this.state.loaded}
              onSubmit={this.onSubmit}
              onChangeIssueInfo={this.onChangeIssueInfo}
              issues={this.props.data.issues}
            />
        );
    }
}


const IssueTableContainerWithData = compose(
    graphql(IssueQuery),
    graphql(IssueUpdate)
)(IssueTable as any);

export default IssueTableContainerWithData;