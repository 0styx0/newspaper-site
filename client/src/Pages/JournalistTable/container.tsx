import * as React from 'react';
import twoDimensionalSorter from '../../helpers/twoDimensionalSorter';
import { compose, graphql } from 'react-apollo';
import { UserQuery, UserUpdate, UserDelete } from '../../graphql/users';
import { User } from './interface.shared';
import JournalistTable from './';
import graphqlErrorNotifier from '../../helpers/graphqlErrorNotifier';

import './index.css';
import { ChangeEvent } from 'react';

export interface State {
    users: User[];
    usersToDelete: Set<string | undefined>;
    idLevelMap: Map<string, number>;
}

export interface Props {
    data: {
        loading: boolean;
        users: User[];
    };
    userUpdate: Function;
    userDelete: Function;
}

/**
 * Creates a table with info about users in it
 *
 * Anyone can see users' name (along with link to profile), number of articles, and total views that person has
 *
 * Logged in users can also see users' levels
 *
 * Logged in users with a level greater than one in the table can also modify that user's level and delete account
 */
export class JournalistTableContainer extends React.Component<Props, State> {

    constructor(props: Props) {
        super(props);

        this.sortInfo = this.sortInfo.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.onDelete = this.onDelete.bind(this);
        this.onIdLevelMap = this.onIdLevelMap.bind(this);

        this.state = {
            users: [],
            usersToDelete: new Set(),
            idLevelMap: new Map()
        };
    }

    /**
     * If data is loaded, sets state.users to props.data.users
     */
    componentWillReceiveProps(props: Props) {

        if (!props.data.users || this.state.users.length > 0) {
            return;
        }

        this.setState({
            users: props.data.users
        });
    }

    /**
     * Sorts by target.value
     */
    sortInfo(event: ChangeEvent<HTMLSelectElement>) {

        const sortBy = (event.target as HTMLSelectElement).value;
        const sortedData = twoDimensionalSorter(this.props.data.users.slice(), sortBy) as User[];

        this.setState({
            users: sortedData
        });
    }

    /**
     * @param e - {htmlInputEvent} whose target.value is a number
     * @param id - id of user
     *
     * Adds an mapping of id -> e.target.value to this.state.idLevelMap
     */
    onIdLevelMap(e: ChangeEvent<HTMLSelectElement>, id: string) {

        const target = e.target as HTMLSelectElement;
        const mapCopy = new Map(this.state.idLevelMap);

        mapCopy.set(id, +target.value);

        this.setState({
            idLevelMap: mapCopy
        });
    }

    /**
     * Adds input value to {@see this.state.usersToDelete} if input is checked, else removes the value
     *
     * @param {htmlInputEvent} - event from html checkbox
     */
    onDelete(e: ChangeEvent<HTMLInputElement>) {

        const target = e.target as HTMLInputElement;

        const copySet = new Set(this.state.usersToDelete);

        target.checked ? copySet.add(target.value) : copySet.delete(target.value);

        this.setState({
            usersToDelete: copySet
        });
    }

    /**
     * @param idLevelMap - map of id to level
     *
     * @return object where level => [ids that mapped to that level]
     *
     * @example convertMapToArrayOfJSON(new Map([['123', 1], ['456', 2], ['789', 1])) =>
     * [
     *   {
     *     level: 1
     *     ids: ['123', '789']
     *   },
     *   {
     *     level: 2,
     *     ids: ['456']
     *   }
     * ]
     */
    convertMapToArrayOfJSON(idLevelMap: Map<string, number>) {

        const data: {level: number; ids: string[]}[] = [];

        Array.from(idLevelMap).forEach(mapping => {

            const [id, level] = mapping;

            const indexOfLevel = data.findIndex(elt => elt.level === level);

            if (indexOfLevel !== -1) {

                data[indexOfLevel].ids.push(id);
            } else {

                data[data.length] = {
                    level,
                    ids: [id]
                };
            }
        });

        return data;
    }

    /**
     * Sends data to server, if there's data to be sent from this.state usersToDelete and/or idLevelMap
     */
    async onSubmit(target: HTMLFormElement) {

        const data = this.convertMapToArrayOfJSON(this.state.idLevelMap);

        const password = (target.querySelector('[name=password]') as HTMLInputElement).value;

        if (data.length > 0) {

            await graphqlErrorNotifier(
                this.props.userUpdate, {
                    variables: {
                        data,
                        password
                    }
                },
                'userUpdate'
            );

            // this line will only fire if sucess since graphqlErrorNotifier throws error otherwise
            const updatedUsers = [...this.state.users].map(user => {

                const updatedLevel = this.state.idLevelMap.get(user.id);
                return Object.assign(JSON.parse(JSON.stringify(user)), {
                    level: updatedLevel || user.level
                });
            });

            this.setState({
                users: updatedUsers,
                idLevelMap: new Map()
            });
        }

        if (this.state.usersToDelete.size > 0) {

            await graphqlErrorNotifier(
                this.props.userDelete, {
                    variables: {
                        ids: [...this.state.usersToDelete],
                        password
                    }
                },
                'userDeleted'
            );

            // this line will only fire if sucess since graphqlErrorNotifier throws error otherwise
            const existingUsers = this.state.users.filter(user => !this.state.usersToDelete.has(user.id));

            this.setState({
                users: existingUsers,
                usersToDelete: new Set()
            });
        }
    }

    render() {

        if (!this.state.users.length) {
            return null;
        }

        return (
            <JournalistTable
              key={this.state.users.map((user: User) => user.id).join(',')}
              users={this.state.users}
              onDelete={this.onDelete}
              onSortInfo={this.sortInfo}
              onSubmit={this.onSubmit}
              onLevelChange={this.onIdLevelMap}
            />
        );
    }
}

const JournalistTableContainerWithData = compose(
    graphql(UserQuery),
    graphql(UserUpdate, {name: 'userUpdate'}),
    graphql(UserDelete, {name: 'userDelete'})
)(JournalistTableContainer);

export default JournalistTableContainerWithData;
