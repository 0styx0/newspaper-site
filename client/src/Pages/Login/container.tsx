import * as React from 'react';
import { setJWT, getJWT } from '../../helpers/jwt';
import LoginForm from './';
import { UserLogin } from '../../graphql/user';
import { graphql, withApollo } from 'react-apollo';
import graphqlErrorNotifier from '../../helpers/graphqlErrorNotifier';

import './index.css';

interface Props {
    history: string[];
    loginUser: ( params:
    {
        variables: {
            username: string,
            password: string
        }
    } ) => Promise<{ data: { login: { jwt: string } }, errors?: {message: string}[] }>;
}

export class LoginFormContainer extends React.Component<Props, {}> {

    constructor() {
        super();

        this.setJWT = this.setJWT.bind(this);
    }

    /**
     * Gets jwt from server and saves to localStorage if valid login info is in form given
     *
     * @param e - even from form with [name=username], [name=password]
     */
    async setJWT(target: HTMLFormElement) {

        const username = (target.querySelector('[name=username]') as HTMLInputElement).value;
        const password = (target.querySelector('[name=password]') as HTMLInputElement).value;

        const { data } = await graphqlErrorNotifier(this.props.loginUser, {
            variables: {
                username,
                password
            }
        });

        setJWT(data.login.jwt);
        const jwt = getJWT();

        if (jwt.level) {

            this.props.history.push('/publish');
        }
        else if (jwt.id) {

            this.props.history.push('/authLogin');
        }
    }

    render() {

        return (
            <LoginForm onLogin={this.setJWT} />
        );
    }
}

const LoginFormContainerWithData = graphql(UserLogin, { name: 'loginUser' })(LoginFormContainer as any);

export default withApollo(LoginFormContainerWithData);
