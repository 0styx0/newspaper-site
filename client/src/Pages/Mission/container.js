import React from 'react';

import httpNotification from '../../helpers/Notification';
import fetchFromApi from '../../helpers/fetchFromApi';

import Mission from './';

export default class MissionContainer extends React.Component {

    constructor() {
        super();

        this.submit = this.submit.bind(this);

        this.state = {
            content: ""
        }
    }

    async componentWillMount() {

        const mission = await fetch("./missionView.html")
                                .then(data => data.text());
        this.setState({
            content: mission
        });
    }

    submit() {


        fetchFromApi("mission", "put", {
                edit: this.state.content
        })
        .then((response) => {

            httpNotification(response.status, response.statusText);
        });

    }

    render() {

        return <Mission
                 content={this.state.content}
                 submit={this.submit}
                 save={(e) => this.setState({content: e.target.innerHTML})}
               />
    }
}