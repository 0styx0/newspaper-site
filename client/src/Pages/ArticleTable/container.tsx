import * as React from 'react';
import { ArticleQuery, ArticleUpdate, ArticleDelete } from '../../graphql/articles';
import { compose, graphql, withApollo } from 'react-apollo';

import ArticleTable from './';

export interface Article {
    tags: {
            all: string[];
        };
    url: string;
    id: string;
    displayOrder: number;
    dateCreated: string;
    views: number;
    author: {
        fullName: string;
        profileLink: string;
    };
}

export interface Issue {
    num: number;
    max: number;
}

interface Props {
    data: {
        loading: boolean;
        issues?: [
            Issue & {
                articles: Article[]
            }
        ]; // will never be more length than 1
    };
    client: {
        query: ( params: { query: typeof ArticleQuery, variables: { issue: number | null; } } ) => Promise<Props>;
    };
    updateArticle: Function;
    deleteArticle: Function;
}

interface State {
    issue: Issue;
    articles: Article[];
    updates: {
        idsToDelete: Set<string>;
        displayOrder: Map<string, number>; // id => info
        tags: Map<string, string[]>;
    };
}

/**
 * Controller of ArticleTable (@see ./index)
 *
 * Gets data from server, then renders ArticleTable with it.
 * This holds listeners for all events that @see ArticleTable can fire
 */
export class ArticleTableContainer extends React.Component<Props, State> {

    constructor() {
        super();

        this.putData = this.putData.bind(this);
        this.convertPropsToState = this.convertPropsToState.bind(this);
        this.onChange = this.onChange.bind(this);
        this.onDelete = this.onDelete.bind(this);
        this.onSubmit = this.onSubmit.bind(this);



        this.state = {
            issue: {} as Issue,
            articles: [],
            updates: {
                idsToDelete: new Set<string>(),
                displayOrder: new Map<string, number>(),
                tags: new Map<string, string[]>()
            }
        };
    }

    /**
     * @param num - articles from what issue to get. If null, get latest
     *
     * Gets data needed for table (@see State.Article interface) and changes page url to modifyArticles/currentIssue
     *
     * Depends on window.location.pathname
     */
    async putData(num: number | null = null) {

        const lastPath = +window.location.pathname.split('/').slice(-1)[0];

        num = (isNaN(lastPath) || ((!isNaN(+num!)) && num)) ? num : lastPath;

        this.props.client.query({
            query: ArticleQuery,
            variables: {
                issue: num
            }
        }).then((data: Props) => {
            this.convertPropsToState(data);
        });

        window.history.pushState(
            {},
            `Issue ${num}`, isNaN(lastPath) ?
              `modifyArticles/${num}` :
             num + ''
        );
    }

    /**
     * Seeds state with graphql props
     */
    componentWillReceiveProps(newProps: Props) {
        this.convertPropsToState(newProps);
    }

    /**
     * Sets `state.issue` and `state.articles` to values extracted from props
     */
    convertPropsToState(props: Props) {

        if (!props.data.issues) {
            return;
        }

        let {max, num} = props.data.issues![0];

        this.setState({
            issue: {
                max,
                num
            },
            articles: props.data.issues![0].articles
        });
    }

    /**
     * Saves changes to state so can be submitted later on. This is for updates only. @see onDelete for deletions
     *
     * @uses `e.target.name` as index of state.updates
     * @uses `e.target.value` as the value
     */
    onChange(e: Event, article: Article) {

        e.stopPropagation();
        e.preventDefault();

        const target = e.target as HTMLInputElement | HTMLSelectElement;

        const stateUpdate = this.state.updates;

        const modifyMapCopy = new Map(stateUpdate[target.name]);
        let value: number | string[];

        if ((target as HTMLSelectElement).selectedOptions) { // for tags

            value = Array.from((target as HTMLSelectElement).selectedOptions)
                    .map(option => option.value)
                    .slice(-3)
                    .sort(); // using sort to make it easier to test, no actual reason besides that
        } else {
            value = +target.value;
        }

        modifyMapCopy.set(article.id, value);

        stateUpdate[target.name] = modifyMapCopy;

        this.setState({
            updates: stateUpdate
        });
    }

    /**
     * Adds/removes user's id from list of users to delete
     *
     * @uses e.target.checked as way to see if should add or remove e.target.value from list
     *
     * @param e {HTMLCheckboxEvent}
     */
    onDelete(e: Event) {

        const stateUpdate = this.state.updates;

        const target = e.target as HTMLInputElement;

        target.checked ? stateUpdate.idsToDelete.add(target.value) : stateUpdate.idsToDelete.delete(target.value);

        this.setState({
            updates: stateUpdate
        });
    }

    /**
     * Sends data from `state.updates` to server
     */
    onSubmit(e: Event) {

        e.stopPropagation();
        e.preventDefault();

        // separating update and delete into functions just to show they're separate stuff
        const submitUpdated = () => {

            const { displayOrder, tags } = this.state.updates;

            if (displayOrder.size + tags.size < 1) {
                return;
            }

            let data: {id: string, displayOrder?: number, tags?: string[]}[] = [...displayOrder].map(mapping => ({

                id: mapping[0],
                displayOrder: mapping[1]
            }));

            [...tags].forEach(mapping => {

                const idIdx = data.findIndex(elt => elt.id === mapping[0]);

                if (idIdx !== -1) {
                    data[idIdx].tags = mapping[1];
                } else {
                    data.push({
                        id: mapping[0],
                        tags: mapping[1]
                    });
                }
            });

            this.props.updateArticle({
                variables: {
                    data
                }
            });
        };

        const submitDeleted = () => {

            if (this.state.updates.idsToDelete.size > 0) {

                this.props.deleteArticle({
                    variables: {
                        ids: [...this.state.updates.idsToDelete]
                    }
                });
            }
        };

        submitUpdated();
        submitDeleted();
    }

    render() {

        if (this.state.articles.length < 1) {
            return null;
        }

        return (
            <ArticleTable
                articles={this.state.articles}
                key={this.state.issue.num}
                issue={this.state.issue}
                onChange={this.onChange}
                onDelete={this.onDelete}
                onUpdate={(e: Event) => this.putData(+(e.target as HTMLInputElement).value)}
                onSubmit={this.onSubmit}
            />
        );
    }
}

const ArticleTableContainerWithData = compose(
    graphql(ArticleQuery, {
        options: {
            variables: {
                issue: +window.location.pathname.split('/').slice(-1)[0]
            }
        }
    }),
    graphql(ArticleUpdate, {name: 'updateArticle'}),
    graphql(ArticleDelete, {name: 'deleteArticle'})
)(ArticleTableContainer as any);

export default withApollo(ArticleTableContainerWithData);