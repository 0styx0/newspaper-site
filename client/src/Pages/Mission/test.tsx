import * as React from 'react';
import { MissionContainer, Props } from './container';
import * as renderer from 'react-test-renderer';
import * as sinon from 'sinon';
import * as casual from 'casual';
import { mount } from 'enzyme';

document.queryCommandSupported = () => true; // used in Editable component

describe('<MissionContainer>', () => {

    /**
     * For use in props.editMission
     *
     * @param newMission - payload
     *
     * @return function editMission used in props
     */
    function fakeEditMission(newMission: string = casual.sentences()) {

        return async (params: { query: Function, variables: { mission: string } }) => ({
            data: {
                editMission: {
                    mission: newMission
                }
            }
        });
    }

    describe('snapshot', () => {

        function snap(canEdit: boolean) {

            const tree = renderer.create(
                <MissionContainer
                    editMission={fakeEditMission()}
                    data={{
                        mission: {
                            mission: casual.sentences(),
                            canEdit
                        }
                    }}
                />
            ).toJSON();

            expect(tree).toMatchSnapshot();
        }

        it('lets user edit when canEdit = true', () => snap(true));

        it('does not let user edit canEdit = false', () => snap(false));
    });

    type editMissionType = Props['editMission'];

    /**
     * @param newMission @see fakeEditMission's newMission
     *
     * @return mounted MissionContainer
     */
    function setup(editMission: editMissionType) {

        const wrapper = mount(
            <MissionContainer
              editMission={editMission || fakeEditMission(casual.sentences())}
              data={{
                  mission: {
                      mission: casual.sentences(),
                      canEdit: true
                  }
              }}
            />
        );

        wrapper.setProps({data: {mission: {
                      mission: casual.sentences(),
                      canEdit: true
                  }}});

        return wrapper;
    }

    describe('editMission', () => {

        it('gets called after submission', () => {

            const spy = sinon.spy();

            const wrapper = setup(function (params) {
                spy();
                return fakeEditMission()(params);
            });

            wrapper.find('button').last().simulate('click');

            expect(spy.called).toBeTruthy();
        });

        it('gets called with correct data (edited mission)', () => {

            const newMission = casual.sentences();

            const wrapper = setup(async (params: { query: Function, variables: { mission: string }}) => {

                expect(params.variables.mission).toBe(newMission);

                return {
                    data: {
                        editMission: {
                            mission: params.variables.mission
                        }
                    }
                };
            });

            const content = wrapper.find('[contentEditable]').first();
            (content.instance() as {} as HTMLDivElement).innerHTML = newMission;
            content.simulate('blur'); // so container can save to state

            wrapper.find('button').last().simulate('click');
        });
    });
});