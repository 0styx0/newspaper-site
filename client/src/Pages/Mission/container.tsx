import * as React from 'react';

import { graphql, withApollo, compose } from 'react-apollo';
import { MissionEdit, MissionQuery } from '../../graphql/mission';

import Mission from './';

export interface Props {
    editMission: (params: { query: typeof MissionEdit, variables: { mission: string } }) => Promise<{
        data: {
            editMission: {
                mission: string
            }
        }
    }>;
    data: {
        mission: {
            mission: string;
            canEdit: boolean;
        }
    };
}

interface State {
    content: string;
}

export class MissionContainer extends React.Component<Props, State> {

    constructor() {
        super();

        this.onSubmit = this.onSubmit.bind(this);

        this.state = {
            content: ''
        };
    }

    /**
     * Gets mission statement
     */
    async componentWillReceiveProps(props: Props) {

        if (props.data.mission.mission) {

            this.setState({
                content: props.data.mission.mission
            });
        }
    }

    /**
     * Sends updates mission to server
     */
    onSubmit() {

        this.props.editMission({
            query: MissionEdit,
            variables: {
                mission: this.state.content
            }
        });
    }

    render() {

        if (!this.state.content) {
            return null;
        }
        
        return (
            <Mission
                content={this.state.content}
                onSubmit={this.onSubmit}
                onSave={(e: Event) => this.setState({content: (e.target as HTMLElement).innerHTML}) as any}
                canEdit={this.props.data.mission.canEdit}
            />
        );
    }
}

const MissionContainerWithData = compose(
    graphql(MissionEdit, {name: 'editMission'}),
    graphql(MissionQuery),
)(MissionContainer as any);

export default withApollo(MissionContainerWithData);
