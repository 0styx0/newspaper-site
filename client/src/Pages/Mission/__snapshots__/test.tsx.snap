// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MissionContainer> snapshot does not let user edit canEdit = false 1`] = `
<div
  className="mission"
  dangerouslySetInnerHTML={
    Object {
      "__html": "",
    }
  }
  onBlur={[Function]}
/>
`;

exports[`<MissionContainer> snapshot lets user edit when canEdit = true 1`] = `
<div>
  <div
    id="buttonContainer"
  >
    <button
      className="bold"
      onClick={[Function]}
      type="button"
    >
      bold
    </button>
    <button
      className="italic"
      onClick={[Function]}
      type="button"
    >
      italic
    </button>
    <button
      className="underline"
      onClick={[Function]}
      type="button"
    >
      underline
    </button>
    <button
      className="heading"
      onClick={[Function]}
      type="button"
    >
      heading
    </button>
    <button
      className="fontSize"
      onClick={[Function]}
      type="button"
    >
      fontSize
    </button>
    <button
      className="insertOrderedList"
      onClick={[Function]}
      type="button"
    >
      insertOrderedList
    </button>
    <button
      className="insertUnorderedList"
      onClick={[Function]}
      type="button"
    >
      insertUnorderedList
    </button>
    <button
      className="justifyFull"
      onClick={[Function]}
      type="button"
    >
      justifyFull
    </button>
    <button
      className="justifyLeft"
      onClick={[Function]}
      type="button"
    >
      justifyLeft
    </button>
    <button
      className="justifyRight"
      onClick={[Function]}
      type="button"
    >
      justifyRight
    </button>
    <button
      className="justifyCenter"
      onClick={[Function]}
      type="button"
    >
      justifyCenter
    </button>
    <button
      className="strikeThrough"
      onClick={[Function]}
      type="button"
    >
      strikeThrough
    </button>
    <button
      className="subscript"
      onClick={[Function]}
      type="button"
    >
      subscript
    </button>
    <button
      className="superscript"
      onClick={[Function]}
      type="button"
    >
      superscript
    </button>
    <button
      className="insertImage"
      onClick={[Function]}
      type="button"
    >
      insertImage
    </button>
    <button
      className="createLink"
      onClick={[Function]}
      type="button"
    >
      createLink
    </button>
    <button
      className="unlink"
      onClick={[Function]}
      type="button"
    >
      unlink
    </button>
    <button
      className="hideFromPreview"
      onClick={[Function]}
      type="button"
    >
      hideFromPreview
    </button>
    <br />
    <button
      onClick={[Function]}
      type="button"
    >
      Submit
    </button>
  </div>
  <div
    className="mission"
    contentEditable={true}
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
    onBlur={[Function]}
  />
</div>
`;
